major update: Python 3, NDB datastore, Flask, pytest

benley opened the pull request:
I apologize for this being such a giant change all at once, but I found it to be impractical to do all this as incremental steps.  My goal was to modernize the code just enough to use Python 3, which required updating the web framework, and Flask made for a pretty good upgrade path from webapp2.  Switching to the NDB datastore was not strictly required, but that turns out to be an easy process.

Unit tests now use pytest with test fixtures from an in-memory datastore library so no external services are needed to urn tests.

Library and tool dependencies have their versions pinned using Nix.  This has no effect on the appengine runtime, but it makes setting up a development environment much easier.

There should be no incompatible changes to the data models, so upgrading an existing site will theoretically be easy.

Some README updates are probably warranted.  Please let me now if you see any major omissions, or if any of the changes I've made are no good; I am open to reworking things as needed.

- [x] python 3 update
- [x] Flask framework to replace webapp2
- [x] Cloud NDB instead of legacy appengine datastore
- [x] Associate slack user IDs with snippets user entities (don't blindly trust slack profile email addresses)
- [x] Email reminders
- [x] Use GCP cloud logging
- [x] http POST instead of GET for modifying state
- [x] Fix unit tests
- [x] hide/unhide/delete users (with javascript enabled)
- [ ] hide/unhide/delete users WITHOUT javascript
- [x] Dependency version pinning with nix for development

benley committed:
Rewrite to use Flask, support Python 3

benley committed:
More python 3 & flask updates

benley committed:
Rewrite to use ndb datastore

benley committed:
Fix python2->3 cmp sorting thing

benley committed:
cleanup app.yaml

benley committed:
Use google cloud logging library

benley committed:
Always use SSL

benley committed:
Fix /admin/update_settings, fix hostname setting

benley committed:
use POST for /admin/update_settings

benley committed:
Swap some RuntimeErrors for http 403 responses

benley committed:
Finally got a working shell.nix for this project

benley committed:
Fix email & chat reminder handlers

benley committed:
Better error handling in update_snippet

benley committed:
2to3 on *_test.py

(they still don't actually WORK)

benley committed:
Progress on slack bot and fixing tests

benley committed:
Authenticate slack users by their slack ID, don't trust emails

The email field in a slack user profile isn't necessarily trustworthy -
some workspaces allow users to edit it themselves.  It also doesn't
necessarily match the user's Google account email.

This change adds a `slack_id` attribute to User entities in the
datastore, and the bot prompts users to manually configure this when
they first use a /snippets command in slack.

We can improve the onboarding UX in the future - I'm sure there is a way
of establishing Google logins from a slack app without copy-and-paste.

benley committed:
Only use cloud logging if we're running on appengine

benley committed:
Fix hide/unhide/delete users (requires javascript)

benley committed:
Fix sending text to hipchat (untested, but probably works)

benley committed:
Add warmup requests handler, set min_instances=1

benley committed:
Prune unused code

benley committed:
Start to appease pyright

benley committed:
Remove eventual-consistency workarounds

Cloud NDB is strongly consistent by default now

benley committed:
Fix most of the tests!

- Use pytest as the runner so we can use its fixture decorators and
  monkeypatch tricks
- Switch to flask's own testing features instead of webtest
- Add an in-memory stub implementation of the ndb API so we don't have
  to involve the datastore emulator binary from the google SDK

benley committed:
Fix the rest of the tests!

benley committed:
Suppress that one deprecation warning

benley committed:
Use time-machine consistently, fix up super() calls

benley committed:
more docs links

benley committed:
Pin nixpkgs with niv, fix entrypoints

benley committed:
Send friday reminder at 1pm Pacific

benley committed:
fix static URLs in local dev, add make typecheck

benley committed:
Use Procfile for local dev with ndb emulator

csilvers commented:
@benley -- I love seeing this, thank you so much for doing it!  This service has needed such an upgrade for a long time.

I'm sorry I haven't reviewed it yet, I've been on a pretty long vacation.  Once I have my head above water again I'll take a look.

In the meantime, can you sign the Khan Academy CLA if you haven't already?  https://www.khanacademy.org/r/cla

csilvers commented on shell.nix:
```diff
@@ -0,0 +1,105 @@
+{ sources ? import ./nix/sources.nix }:
```
I don't know anything about nix, so I can't really review this part.  But I'm wondering if this file can also live in the nix/ directory -- I'm guessing not.

benley replied:
This one file is most useful if it's in the root of the repository.  You activate it by running `nix-shell`, which looks for shell.nix in your current directory.  It's also not strictly necessary to include any of the nix bits in this pull request, but I found it very helpful to use nix for pinning all the external dependencies.

csilvers replied:
OK, that's fine then.

Which reminds me: this PR should also update the README to give instructions on how to run the server manually.  It looks like you'll need to install nix on your machine first?  And then run what?

And also instructions on how to update the nix configs if that's something that might ever be needed.  Like, if we add a feature that requires a new OS package to be installed, what do we do to get that to work in nix-land?  Ideally something that's understandable even by people who don't know anything about nix.

csilvers commented on slacklib_test.py:
```diff
@@ -1,70 +1,76 @@
-
+# The fictional day for these tests Wednesday, July 29, 2015
+@time_machine.travel(datetime.datetime(2015, 7, 29, tzinfo=ZoneInfo("UTC")))
 class SlashCommandTest(unittest.TestCase):
```
I'm confused about this because you're using pytest now, right?  Does pytest support unittest.TestCase classes?  Did you have a desire to move this to a more idiomatic pytest format?  (Not in this diff, of course, but you could add a TODO if you wanted to.)

benley replied:
pytest does support unittest.TestCase classes!  My general plan was to gradually transition the tests to a more idiomatic pytest style, but not all in this diff.

csilvers replied:
That sounds great.  I'd just put in TODOs in the places you want to transition.  But it's up to you.

csilvers commented on snippets.py:
```diff
@@ -9,43 +9,54 @@
 __author__ = 'Craig Silverstein <csilvers@khanacademy.org>'
 
 import datetime
+from functools import cmp_to_key
```
Style nit: khan academy style is to prefer to import entire modules, rather than symbols from modules.  So just import `functools` and use `functools.cmp_to_key` below.

csilvers commented on snippets.py:
```diff
@@ -371,311 +380,318 @@ def post(self):
+        try:
+            week = datetime.datetime.strptime(week_string, '%m-%d-%Y').date()
+        except (ValueError, TypeError):
+            return flask.make_response({"status": 400, "message": "Invalid week. Expected format: MM-DD-YYYY"}, 400)
```
Nice!

csilvers commented on snippets.py:
```diff
@@ -371,311 +380,318 @@ def post(self):
+    # NOTE: this will break if you explicitly set the key when creating the
+    #       user entity!
+    #       See https://groups.google.com/g/google-appengine/c/Tm8NDWIvc70
+    if user.key and user.key.id():
+        is_new_user = False
+    else:
+        is_new_user = True
```
```suggestion
    is_new_user = bool(user.key and user.key.id())
```

benley replied:
oh nice, that is much better.

csilvers commented on snippets.py:
```diff
@@ -371,311 +380,318 @@ def post(self):
+    # We have to use 'cmp' here since we want ascending in the
+    # primary key and descending in the secondary key, sometimes.
+    if sort_by == 'email':
+        user_data.sort(key=cmp_to_key(lambda x, y: cmp(x[0], y[0])))
```
Do you need all this cmp_to_key/cmp stuff?  Couldn't you just do `key=lambda x: x[0]`?

benley replied:
You're right, it looks like that accomplishes the same thing.  Fixing!

csilvers commented on snippets.py:
```diff
@@ -725,80 +742,78 @@ def _maybe_send_snippets_mail(to, subject, template_path, template_values):
+    return "OK", 200, {}
+
+
+app.add_url_rule("/slack", view_func=slacklib.slash_command_handler,
+                 methods=["POST"])
+app.add_url_rule("/admin/test_send_to_hipchat",
+                 view_func=hipchatlib.test_send_to_hipchat_handler)
```
This is a bit awkward.  Would it make more sense to put the app in a separate app.py file, and let slacklib and hipchatlib import app.py and use the normal `@app.route`?

csilvers replied:
I wouldn't spend a huge amount of time on this, just something to think about.

benley replied:
That turned out to be nice and simple; done.

csilvers commented on templates/app_settings.html:
```diff
@@ -9,7 +9,7 @@ <h1>Manage Users</h1>
 
 
-<form action="/admin/update_settings" method="get" class="user-settings">
+<form action="/admin/update_settings" method="post" class="user-settings">
```
I wonder what I was thinking, making this a GET originally.

csilvers reviewed (changes requested):
This is an amazing piece of work!  Thank you so much for taking the time to do it.  The new code looks really good to me.

I do have to admit, though, that I'm somewhat uncomfortable having this mondo PR that changes everything at once.  Now that you've made the work to convert everything, I'm wondering if maybe there is a way to split some things into a separate PR.  For instance, could the changes to add slack_id as a setting, and all the changes that come with that, be its own PR?  What about the change to move to flask?

I know it's more work given that you already have this PR, and I'm on the fence about how much of a difference it would make.  Let me know what you think.

"Request changes" for a bunch of small things -- nothing major, except maybe the breaking change in how we handle slack-ids.  (Which is one reason it might be nice to separate that part, at least, out to its own PR.)  I'm excited to see this land!

benley commented on app.yaml:
```diff
@@ -1,7 +1,15 @@
 default_expiration: "365d"
+app_engine_apis: true
+
+# Slack has a very short timeout on API callbacks, so we need to keep at least 1
+# process running for it
+automatic_scaling:
+  min_instances: 1
```
I never ran this until recently so I'm not sure what the earlier behavior was.  That said, I have found that min_instances=1 still fits within the appengine free tier, so it should still be possible to run without billing enabled.  We could also omit that from the defaults in app.yaml if it's not appropriate for all users.

csilvers replied:
Gotcha.  If it fits in the free tier let's keep it for now.

benley commented on conftest.py:
```diff
@@ -0,0 +1,55 @@
+    monkeypatch.setenv("DATASTORE_DATASET", "datastore-stub-test")
+
+
+# @pytest.fixture
+# def snippets_app():
+#     app = snippets.app
+#     # any setup stuff can go here, e.g.
+#     # app.config.update({
+#     #     "TESTING": True,
+#     # })
+#     yield app
+#     # cleanup / reset resources here
+#
+#
+# @pytest.fixture
+# def snippets_client(snippets_app):
+#     return snippets_app.test_client()
+#
+#
+# @pytest.fixture
+# def snippets_runner(snippets_app):
+#     return snippets_app.test_cli_runner()
```
I think this was stuff I was considering for a future iteration; I will remove the commented out bits for now.

benley commented on hipchatlib.py:
```diff
@@ -9,18 +9,19 @@
 import models
 
 
 def _make_hipchat_api_call(post_dict_with_secret_token):
```
ooooh good point, it looks like it does not!  Should I go ahead and remove the hipchat module?

csilvers replied:
Yeah, let's do that!  But as a separate PR.  Then you can rebase this PR on top of that one.

benley replied:
here we go: https://github.com/Khan/snippets/pull/36

csilvers replied:
Merged, feel free to rebase this one.

benley commented on snippets.py:
```diff
@@ -371,311 +380,318 @@ def post(self):
-            return
-
-        email = self.request.get('u', _current_user_email())
+            return flask.make_response({"status": 401, "message": "not logged in"}, 401)
```
My understanding of the codes is that 403 indicates "I know who you are, and this is forbidden" whereas 401 means "You are not logged in, please try again with credentials"

benley commented on snippets.py:
```diff
@@ -371,311 +380,318 @@ def post(self):
 
-        email = self.request.get('u', _current_user_email())
-        self.redirect("/?msg=Snippet+saved&u=%s" % urllib.quote(email))
+    else:  # flask.request.method == "GET"
```
Agreed, I can make that change.  I was considering removing the GET path entirely here, but I don't know if there are existing users who rely on it for integrations.

benley commented on templates/settings.html:
```diff
@@ -27,6 +27,11 @@ <h1>User settings for {{username}}</h1>
 </div>
 
+<div class="user-settings-block">
+  <label class="user-settings-label" for="slack-id">Slack User ID</label>
```
This is actually an opaque identifier that slack assigns (e.g. `U01CT7ABT0K`), and you're right that it warrants an explanation!  If I recall correctly I had some workflow where the slack bot informs you of your userid and tells you where to put it in the app, but there should be inline documentation here.  Will do.

benley commented on slacklib.py:
```diff
@@ -72,40 +66,6 @@ def _web_api(api_method, payload):
     return reply
 
 
-def _get_user_email(uid):
-    """Retrieve the email address for a specific userid from the Slack Web API.
-
-    Raises ValueError if could not be retrieved.
-    """
-    reply = _web_api('users.info', {'user': uid})  # possible ValueError
-    email = reply.get('user', {}).get('profile', {}).get('email', None)
-    if email is None:
-        raise ValueError('Slack user profile did not have email')
-    return email
-
-
-def _get_user_email_cached(uid, force_refresh=False):
-    """Retrieve the email address for a specific user id, with a cache.
-
-    Results are stored in memcache for up to a day.
-
-    If force_refresh parameter is specified, cached data will be refreshed.
-
-    Raises ValueError if could not be retrieved.
-    """
-    key = 'slack_profile_email_' + uid
-    cached_data = memcache.get(key)
-    if (cached_data is None) or force_refresh:
-        logging.debug("cache miss/refresh for slack email lookup %s", uid)
-        email = _get_user_email(uid)  # possible ValueError
-        if not memcache.set(key=key, value=email, time=86400):
-            logging.error('memcache set failed!')
-        return email
-    else:
-        logging.debug("cache hit for slack email lookup %s", uid)
-        return cached_data
```
I ran into an issue at my company where slack user emails don't always match the user's google workspace email, which made me realize there is also a security concern here.  From a commit message (https://github.com/Khan/snippets/pull/35/commits/fa6ff22ebc4cb0d8b784edacdbc005d1df2189d3): 

> **Authenticate slack users by their slack ID, don't trust emails**
> The email field in a slack user profile isn't necessarily trustworthy -
some workspaces allow users to edit it themselves.  It also doesn't
necessarily match the user's Google account email.
>
> This change adds a `slack_id` attribute to User entities in the
datastore, and the bot prompts users to manually configure this when
they first use a /snippets command in slack.
>
> We can improve the onboarding UX in the future - I'm sure there is a way
of establishing Google logins from a slack app without copy-and-paste.

csilvers replied:
OK, I think that's a reasonable change to make, but should definitely be a separate PR.  I'm ok with you doing that PR first (before this one), or after, whatever is easier for you.

csilvers replied:
We'll also have to think about backward compatibility: how we don't break existing users using this feature when we upgrade.

benley replied:
What if we add a toggle in AdminSettings between the old and new behavior?

> - [x] **Trust Slack user email addresses?**
> If enabled, the Snippets server will assume that a Slack user's email address can be safely mapped to a Google account with the same address.  Disable this if your users may have mismatched email addresses between Slack and Google, or if your Slack workspace permits users to arbitrarily edit their own email address field.

We could start lazily populating the `User.slack_id` property when the toggle is enabled, which would eliminate the need for memcache.  And if the toggle is disabled, users would need to manually associate their Slack & Google accounts with the bot's instructions:

> *&lt;user&gt;* /snippets list
> *&lt;SnippetsBot&gt;* You don't seem to be logged in!  Please configure your Slack user ID (which is `U01CT7ABH0R`) in the snippets server's settings page: https://snippets.example.com/settings

csilvers replied:
Yeah, I think adding another setting for this is a fine solution!  All the more reason to do it in a separate PR though :-)

benley replied:
Sounds good, I'll open another one to do that.

benley commented:
> Thank you for the detailed review! I also just got back from vacation and am gradually catching up, so I will do my best to finish responding to comments in the next day or so.

Okay actually it will be another few days before I can actually spend time on this.  But it is not abandoned, I'm coming back to finish when I can.

csilvers commented:
No rush at all!

benley commented on slacklib.py:
```diff
@@ -240,7 +205,7 @@ def command_list(user_email):
         "*Your snippets for the week so far:*\n" +
         _format_snippet_items(items)
-    )
+)
```
Oops, fixing!

benley committed:
Code review feedback, round 1

benley committed:
Whoops, restore the missing wrap_wsgi_app

The app doesn't work on actual appengine without this!

benley commented on snippets.py:
```diff
@@ -371,311 +380,318 @@ def post(self):
+    if sort_by == 'email':
+        user_data.sort(key=cmp_to_key(lambda x, y: cmp(x[0], y[0])))
+    elif sort_by == 'creation_time':
+        user_data.sort(key=cmp_to_key(
```
hmmm this part won't actually work because `-x[0]` is trying to negate a string.

benley replied:
but this works:
```py
        user_data.sort(key=lambda x: x[0])
        user_data.sort(key=lambda x: x[2] or datetime.datetime.min, reverse=True)
```

It's slower, but we probably aren't dealing with enough data for that to really matter.

benley committed:
Simplify sort_by method, drop cmp_to_key

as suggested in python docs:
https://docs.python.org/3/howto/sorting.html#sort-stability-and-complex-sorts

benley committed:
Replace _TODAY_FN with datetime.datetime.now

We're using time_machine to handle datetime in tests now.

benley commented on slacklib.py:
```diff
@@ -43,9 +39,6 @@
 # This allows mocking in a different day, for testing.
 _TODAY_FN = datetime.datetime.now
```
Done, might as well go for it.

csilvers commented on conftest.py:
```diff
@@ -26,30 +26,9 @@ def ndb_context(init_ndb_env_vars):
-    Initializing an ndb Client in a test env requires some environment variables to be set
-    For now, these are just garbage values intended to give the library _something_
-    (we don't expect them to actually work yet)
+    Initializing an ndb Client in a test env requires some environment variables
```
```suggestion
    """Set envvars on test ndb client.
    
    Initializing an ndb Client in a test env requires some environment variables
```

csilvers commented on snippets.py:
```diff
@@ -198,7 +194,8 @@ def user_page_handler():
     snippets = util.fill_in_missing_snippets(snippets, user,
-                                             user_email, _TODAY_FN())
+                                             user_email,
+                                             datetime.datetime.now())
```
You may find it nicer to just set `now = datetime.datetime.now()` at the top of this function and use it everywhere.

benley replied:
You're right, that is much nicer.

csilvers commented on snippets.py:
```diff
@@ -350,79 +348,89 @@ def update_snippet(email: str,
-def update_snippet_handler():
+@app.route("/update_snippet", methods=["POST"])
+def update_snippet_handler_post():
+    """handle ajax updates via POST
```
```suggestion
    """Handle ajax updates via POST.
```
(total pep8 nit.)

csilvers commented on snippets.py:
```diff
@@ -350,79 +348,89 @@ def update_snippet(email: str,
-            return _login_page(flask.request)
+@app.route("/update_snippet", methods=["GET"])
+def update_snippet_handler_get():
 
```
(Another nit: no blank line here.)

csilvers commented on snippets.py:
```diff
@@ -659,20 +659,14 @@ def admin_manage_users_handler():
-                               1000 if y[3] is None else y[3])
-                          or cmp(x[0], y[0]))))
+        user_data.sort(key=lambda x: x[0])
+        user_data.sort(key=lambda x: 1000 if x[3] is None else x[3], reverse=True)
```
This one you could do:
```
user_data.sort(key=lambda x: (-1000 if x[3] is None else -x[3], x[0]))
```
though maybe you want to keep it the way it is for parallelism.

benley replied:
Yeah, I only did it this way to be consistent with the creation_time branch.  Up to you if you'd prefer the single-pass style.

csilvers commented on models.py:
```diff
@@ -21,56 +21,57 @@
-    default_private = db.BooleanProperty(default=False)   # new-user default
-    default_markdown = db.BooleanProperty(default=True)   # new-user default
-    default_email = db.BooleanProperty(default=True)      # new-user default
+    domains = ndb.StringProperty(repeated=True)
```
This still needs to be done.

benley replied:
I need to figure out why I changed this in the first place; iirc there was some kind of bootstrapping issue.

benley replied:
Ohhhhh, it's because fields cannot have both `repeated=True` and `required=True`.  Which option would you prefer?
- StringProperty with required=True (as before), parsed into a list at runtime
- (current) StringProperty with repeated=True, and we can make it effectively required in our code
- JsonProperty with required=True and a validator function to guarantee a list of strings

csilvers replied:
Gotcha.  I think what you have now is good then.

The important thing to me is backward-compat.  As long as db.StringListProperty is wire-compatible with ndb.StringProperty(repeated=True) -- which I think it is -- then this is a good change by me!

csilvers reviewed (changes requested):
Lots of good progress!  I couldn't tell if you were ready for another look yet or not, but I did one anyway.  There are still a few things that (I think!) remain to be done so I still have "request changes".

1. Update the README to give new instructions, esp. for using nix
2. Add back in `required=True` for `domains`
3. Keep `_get_user_email` in slacklib.py, for now.  I'm ok with getting rid of it, but it should be its own PR because it's going to be a bit tricky to roll out

I think that's it!

benley committed:
Merge remote-tracking branch 'origin/master' into python3-flask-ndb

benley committed:
Use flask blueprints to register /slack

benley committed:
pep8 nits

benley committed:
Call datetime.now() only once per function

benley committed:
Another pep8 nit

benley commented:
I know there are still a few things left to address - I will get there soon :)

csilvers commented:
> I know there are still a few things left to address - I will get there soon :)

Just to over-communicate: I'm waiting for you to give the signal before I review this again.  No rush at all, just don't want you to be waiting on me while I'm waiting on you. :-)

benley commented:
> > I know there are still a few things left to address - I will get there soon :)
> 
> Just to over-communicate: I'm waiting for you to give the signal before I review this again. No rush at all, just don't want you to be waiting on me while I'm waiting on you. :-)

Noted!  I will be able to continue work on this in about a week, I think.  Thank you for your patience.

benley commented:
Sorry this is still unfinished - I just wanted to say it's not abandoned, I do still intend to come back and finish this PR in the near future.

csilvers commented:
Thanks for the update!  No rush.
